iceoryx_examples/icediscovery/iox_find_service.cpp

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_hoofs/posix_wrapper/signal_watcher.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+#include "iceoryx_posh/runtime/service_discovery.hpp"
+
+#include <iostream>
+
+constexpr char APP_NAME[] = "iox-find-service";
+
+void printSearchResult(const iox::cxx::expected<iox::runtime::ServiceContainer, iox::runtime::FindServiceError>& res)
+{
+    res.and_then([](auto& serviceContainer) {
+           for (auto& service : serviceContainer)
+           {
+               std::cout << "- " << service << std::endl;
+           }
+           std::cout << std::endl;
+       })
+        .or_else([](auto& error) {
+            std::cerr << "findService() call failed with: " << static_cast<uint64_t>(error) << std::endl;
+        });
+}
+
+int main()
+{
+    iox::runtime::PoshRuntime::initRuntime(APP_NAME);
+
+    iox::runtime::ServiceDiscovery serviceDiscovery;
+
+    while (!iox::posix::hasTerminationRequested())
+    {
+        std::cout << "=========================================" << std::endl;
+
+        std::cout << "Searched for {'Radar', 'FrontLeft', 'SequenceCounter'} and found the following events: "
+                  << std::endl;
+        auto searchResult = serviceDiscovery.findService(iox::capro::IdString_t{"Radar"},
+                                                         iox::capro::IdString_t{"FrontLeft"},
+                                                         iox::capro::IdString_t{"SequenceCounter"});
+        printSearchResult(searchResult);
+
+        std::cout << "Searched for {'Radar', *, *} and found the following events: " << std::endl;
+        searchResult =
+            serviceDiscovery.findService(iox::capro::IdString_t{"Radar"}, iox::capro::Wildcard, iox::capro::Wildcard);
+        printSearchResult(searchResult);
+
```

Comment:
I would love to see a `findService` variation with the following signature:
```cpp
void findService(const cxx::optional<capro::IdString_t>& service, const cxx::optional<capro::IdString_t>& instance, const cxx::optional<capro::IdString_t>& event, const cxx::function_ref<void(ServiceDescription&)> & callback);
```

We already have the problem on platforms like Windows & QNX that the stack size is increasing and this would solve this issue and would enable the developer to have a more memory efficient alternative.
An example could look like:

```cpp
serviceDiscovery.findService(iox::capro::IdString_t{"Radar"},
                                                         iox::capro::Wildcard,
                                                         iox::capro::Wildcard, [](auto s) { std::cout << s << std::endl; );
```

Even in the implementation it would not cause any overhead from the lines of code. One could implement such a thing very easily like:

```cpp
searchResult_t findService(const cxx::optional<capro::IdString_t>& service, const cxx::optional<capro::IdString_t>& instance, const cxx::optional<capro::IdString_t>& event) {
  searchResult_t result;
  findService(service, instance, event, [&](auto & service){ result.emplace_back(service); });
  return result;
}
```

Could you please therefore add this to the API and the example!

Reply:
@elfenpiff Already planned as stated [here](https://github.com/eclipse-iceoryx/iceoryx/pull/1083#issuecomment-1034016136) ;)

Reply:
I agree with @elfenpiff but I would step by step and maybe not on this PR but on a follow-up one. "Make it work, make it beautiful, make it fast"

Reply:
@mossmaurice As I stated [here](https://github.com/eclipse-iceoryx/iceoryx/pull/1083#issuecomment-1034016136), this will be part of a follow-up PR.

Reply:
@FerdinandSpitzschnueffler A lambda API that propagates down to even the `ServiceRegistry` is planned. And it is well-known that we can implement the current API in terms of this API. One problem at a time ...

Besides another nice alternative is working with views, but this is out of scope as it creates issues with storing callables and one needs to be careful with lifetime then.